src/domain/core/src/repos/metadata_chain_visitors.rs

```diff
@@ -0,0 +1,404 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::error::Error;
+use std::marker::PhantomData;
+
+use internal_error::InternalError;
+use opendatafabric::{
+    AddData,
+    AsTypedBlock,
+    ExecuteTransform,
+    IntoDataStreamBlock,
+    MetadataBlock,
+    MetadataBlockDataStream,
+    MetadataBlockTyped,
+    MetadataEvent,
+    MetadataEventDataStream,
+    MetadataEventTypeFlags as Flag,
+    Multihash,
+    Seed,
+    SetAttachments,
+    SetDataSchema,
+    SetInfo,
+    SetLicense,
+    SetPollingSource,
+    SetTransform,
+    SetVocab,
+    SourceState,
+    VariantOf,
+};
+
+use crate::{
+    HashedMetadataBlockRef,
+    IterBlocksError,
+    MetadataChainVisitor,
+    MetadataChainVisitorHolder,
+    MetadataChainVisitorHolderFactory,
+    MetadataVisitorDecision as Decision,
+};
+
+///////////////////////////////////////////////////////////////////////////////
+
+// TODO: use Infallible/!
+type VisitorError = IterBlocksError;
```

Comment:
I played with `Infallible` (from stable) and `!` (from unstable) for a long time, in general I had a prototype, but it was very verbose, as in some (almost all situations), we had to duplicate types for method calls.

From the point of view of type purity, it's a compromise solution, I realize that -- in result, to make the final (caller) code simpler, I decided to use `IterBlocksError`.